sdk/core/Azure.Core/src/Shared/AsyncLockWithValue.cs

```diff
@@ -108,73 +154,107 @@ private void SetValue(T value)
 
             while (waiters.Count > 0)
             {
-                waiters.Dequeue().TrySetResult(new Lock(value));
+                waiters.Dequeue().TrySetResult(new LockOrValue(value));
             }
         }
 
         /// <summary>
         /// Release the lock and allow next waiter acquire it
         /// </summary>
-        private void Reset()
+        private void Reset(in long lockIndex)
         {
-            TaskCompletionSource<Lock> nextWaiter = UnlockOrGetNextWaiter();
-            while (nextWaiter != default && !nextWaiter.TrySetResult(new Lock(this)))
+            UnlockOrGetNextWaiter(lockIndex, out var nextWaiter);
+            while (nextWaiter != default && !nextWaiter.TrySetResult(new LockOrValue(this, unchecked(lockIndex + 1))))
             {
-                nextWaiter = UnlockOrGetNextWaiter();
+                UnlockOrGetNextWaiter(lockIndex, out nextWaiter);
             }
         }
 
-        private TaskCompletionSource<Lock> UnlockOrGetNextWaiter()
+        private void UnlockOrGetNextWaiter(in long lockIndex, out TaskCompletionSource<LockOrValue>? nextWaiter)
         {
             lock (_syncObj)
             {
-                if (!_isLocked)
+                nextWaiter = default;
+                // If lock isn't acquired, just return
+                if (!_isLocked || lockIndex != _index)
                 {
-                    return default;
+                    return;
                 }
 
+                _index = unchecked(lockIndex + 1);
+
+                // If lock was acquired, but there are no waiters, unlock and return
                 if (_waiters == default)
                 {
                     _isLocked = false;
-                    return default;
+                    return;
                 }
 
+                // Find the next waiter
                 while (_waiters.Count > 0)
                 {
-                    var nextWaiter = _waiters.Dequeue();
+                    nextWaiter = _waiters.Dequeue();
                     if (!nextWaiter.Task.IsCompleted)
                     {
                         // Return the waiter only if it wasn't canceled already
-                        return nextWaiter;
+                        return;
                     }
                 }
 
+                // If no next waiter has been found, unlock and return
                 _isLocked = false;
-                return default;
             }
         }
 
-        public readonly struct Lock : IDisposable
+        public readonly struct LockOrValue : IDisposable
```

Comment:
It seems we're changing the functionality of this type.  Is there still a use case for just the lock, and would it make sense to instead introduce a second LockOrValue type for just the use cases where value is needed?

Reply:
No, functionality remains the same. It has always been "lock or value" depending on either value is cached or not. I've just thought that `LockOrValue` name is better.